secscanner2junit/sast.py

```diff
@@ -64,7 +64,7 @@ def parse_findings(self, finding, time):
 
         f_type = finding['identifiers'][0]['name']
         if properties['name']:
-            tc = TestCase(name=properties['name'], classname=self.p_type, file=properties['file'], elapsed_sec=time, line=properties['start_line'])
+            tc = TestCase(name=properties['name'] + " " properties['id'] , classname=self.p_type, file=properties['file'], elapsed_sec=time, line=properties['start_line'])
```

Comment:
1. fixed

Reply:
2. I don't understand, IMO every item in `vulnerabilities` table has unique id:
```json
   {
      "id": "db914ce5737b49650ae650fc3b0fe38a531eadd8ea780f48a013419c4adec7f0",
      "category": "sast",
      "name": "Found Spring endpoint",
      "message": "Found Spring endpoint",
      "description": "pl.........PingController is a Spring endpoint (Controller)",
      "cve": "21254b1dfdebd6b8bbd05e4ed8a960c3:SPRING_ENDPOINT:src/main/java/pl/com/softnet/example/springbootsoftnetexample/PingController.java:23",
      "severity": "Low",
      "confidence": "Low",
      "scanner": {
        "id": "find_sec_bugs",
        "name": "Find Security Bugs"
      },
      "location": {
        "file": "src/main/java/pl/........./PingController.java",
        "start_line": 23,
        "class": "pl..........PingController",
        "method": "ping"
      },
      "identifiers": [
        {
          "type": "find_sec_bugs_type",
          "name": "Find Security Bugs-SPRING_ENDPOINT",
          "value": "SPRING_ENDPOINT",
          "url": "https://find-sec-bugs.github.io/bugs.htm#SPRING_ENDPOINT"
        }
      ]
    },
    {
      "id": "f4e1ee2a65c5d8837cfe6e3b16fc368f23462596f41ca15b182625a259a58baf",
      "category": "sast",
      "name": "Found Spring endpoint",
      "message": "Found Spring endpoint",
      "description": "pl........FakeErrorController is a Spring endpoint (Controller)",
      "cve": "62a35767e47f86da1958c888ab0ddb98:SPRING_ENDPOINT:src/main/java/pl/com/softnet/example/springbootsoftnetexample/FakeErrorController.java:16",
      "severity": "Low",
      "confidence": "Low",
      "scanner": {
        "id": "find_sec_bugs",
        "name": "Find Security Bugs"
      },
      "location": {
        "file": "src/main/java/pl/......./FakeErrorController.java",
        "start_line": 16,
        "class": "pl............FakeErrorController",
        "method": "getDomainError"
      },
      "identifiers": [
        {
          "type": "find_sec_bugs_type",
          "name": "Find Security Bugs-SPRING_ENDPOINT",
          "value": "SPRING_ENDPOINT",
          "url": "https://find-sec-bugs.github.io/bugs.htm#SPRING_ENDPOINT"
        }
      ]
    },
```

Reply:
As per my understanding of the SAST report format, the 'id' field is not necessarily given. (See [here](https://gitlab.com/gitlab-org/security-products/security-report-schemas/-/blob/master/dist/sast-report-format.json#L550-L556))

And even though each vulnerability contains an 'identifiers' field which is always given, if you look at your example, you'll see, that the entries at identical for both vulnerabilities...

Reply:
(See [here](https://gitlab.com/gitlab-org/security-products/security-report-schemas/-/blob/master/dist/sast-report-format.json#L550-L556)) - yes this is a problem, but i have never seen entry without id. Maybe it is some legacy that it is not required but always included ?

The problem is that, this two entries are merged into one on MR view and it would be useful to know all occurrences of vulnerability.

Reply:
I found the data provided by different SAST scanners [here](https://docs.gitlab.com/ee/user/application_security/sast/analyzers.html#data-provided-by-analyzers). It seems like only three scanners don't provide this kind of `id`.

However, I would prefer to not dismiss these cases. What do you think about the following:

```python
try:
   finding_id = properties['id']
except KeyError:
   finding_id = str(random.randrange(1, 10000000))
if properties['name']:
   tc = TestCase(name=properties['name'] + " " + finding_id , classname=self.p_type, file=properties['file'], elapsed_sec=time, line=properties['start_line'])
   tc = TestCase(name=properties['name'] + " " + finding_id , classname=self.p_type, file=properties['file'], elapsed_sec=time, line=properties['start_line'])
   else:
      tc = TestCase(name=f_type, classname=self.p_type, file=properties['file'], elapsed_sec=time, line=properties['start_line'])
```

This way, the probability of two testcases having the same name is extremely small, and even if so, on the next run they would definitely be listed as two distinct testcases.

Reply:
You are my HERO 🥇 best proposition ever !!!